# Enriching data
#### 1. Adding data from www.themoviedb.org (TMDB)

In this section we are going to be using the API of www.themoviedb.org (TMDB) to populate the existing dataframe with more information about the movie. The script is going to query the website with the movie name and match with the movie, where the release date is the closest. If no match is found the code skips the movie and goes to the next.

This 1. step is most importantly adding the TMDB_ID which is a unique identifier for the movie which can be used to further enrich the data.

This script was run separately in another file in /scratch/add_TMDB_movie_metadata.py as it needed to be run several times since the API was limited by 40 requests/s.
 


In [2]:
import tmdbsimple as tmdb # Wrapper library for the API of themoviedb.org  (TMDB)
from tqdm import tqdm # Progress bar for the script
import pandas as pd
from datetime import datetime # Used for comparing movie release dates
from dotenv import load_dotenv # Makes keeping the API-key as local environment file simpler
import os # Used for loading the .env file

load_dotenv() # Loads .env files aka. the TMDB_API_KEY

headers_movie_metadata = ["Wikipedia Movie ID", "Freebase Movie ID", "Movie name", "Movie release date",
                          "Movie box office revenue", "Movie runtime", "Movie languages", "Movie countries",
                          "Movie genres"]
movie_metadata = pd.read_csv('data/movie.metadata.tsv', sep="\t", names=headers_movie_metadata)


# Load API key
TMDB_API_KEY = os.environ.get("TMDB_API_KEY")
tmdb.API_KEY = TMDB_API_KEY
tmdb.REQUESTS_TIMEOUT = 5  # Seconds, for both connect and read

# Create a list to save progress
saved_progress = []

# Determine where to resume
start_index = 28000

# Progress file that can be used to resume
#saved_progress = pd.read_json('progress.json')['index'].tolist()
#start_index = saved_progress[-1] + 1  # Start from the next index

# Create a DataFrame to store the data
movie_metadata_TMDB = movie_metadata.copy()

for index, row in tqdm(movie_metadata_TMDB.iterrows(), total=len(movie_metadata_TMDB), desc="Processing"): # Wraps for loop in progress bar.
    
    # Skip previously processed indices - Commented out in favor of manual start_index
    #if index in saved_progress:
    #    continue
    if index < start_index:
        continue
    try:
        if not pd.isna(row["Movie release date"]):
            search = tmdb.Search()
            response = search.movie(query=row["Movie name"])

            # Convert dataframe release date to datetime
            movie_release_date_str = row["Movie release date"]
            if len(movie_release_date_str) == 4:  # Handle "YYYY" format
                movie_release_date = datetime.strptime(movie_release_date_str, "%Y").date()
            elif len(movie_release_date_str) == 7:  # Handle "YYYY-DD" format
                movie_release_date = datetime.strptime(movie_release_date_str, "%Y-%m").date()
            else:  # Assume it's in the format "YYYY-MM-DD"
                movie_release_date = datetime.strptime(movie_release_date_str, "%Y-%m-%d").date()

            # Handle if release_date is empty.
            date_list_converted = [datetime.strptime(each_date['release_date'], "%Y-%m-%d").date() for each_date in
                                   search.results if each_date.get('release_date')]

            # Create list of differences in time
            differences = [abs(movie_release_date - each_date) for each_date in date_list_converted]
            
            # If differences are empty = skip
            if not differences:
                continue
            minimum_index = differences.index(min(differences))  # Index of the closest match
            match = search.results[minimum_index]
            # print(f"Closest match: {match['title']} (Release Date: {match['release_date']})")

            # Add info in dataframe about the movie
            movie_metadata_TMDB.loc[index, 'TMDB_id'] = match['id']
            movie_metadata_TMDB.loc[index, 'TMDB_original_language'] = match['original_language']
            movie_metadata_TMDB.loc[index, 'TMDB_original_title'] = match['original_title']
            movie_metadata_TMDB.loc[index, 'TMDB_overview'] = match['overview']
            movie_metadata_TMDB.loc[index, 'TMDB_popularity'] = match['popularity']
            movie_metadata_TMDB.loc[index, 'TMDB_release_date'] = match['release_date']
            movie_metadata_TMDB.loc[index, 'TMDB_title'] = match['title']
            movie_metadata_TMDB.loc[index, 'TMDB_vote_average'] = match['vote_average']
            movie_metadata_TMDB.loc[index, 'TMDB_vote_count'] = match['vote_count']
            
            # Save the index as progress
            saved_progress.append(index)

            # Save progress periodically (in case of interruption)
            if index % 50 == 0:
                progress_df = pd.DataFrame({'index': saved_progress})
                progress_df.to_json('progress.json')
                movie_metadata_TMDB.to_csv('movie_metadata_TMDB.csv', index=False)

    except Exception as e:
        print(f"Error at index {index}: {str(e)}")

# Save final progress
progress_df = pd.DataFrame({'index': saved_progress})
progress_df.to_json('progress.json')

# Save your final DataFrame
movie_metadata_TMDB.to_csv('modified_data/movie_metadata_TMDB.csv')

Processing: 100%|██████████| 81741/81741 [00:03<00:00, 20657.45it/s]


In [3]:
# Comparing how many rows were populated in the enriching.
import pandas as pd
df = pd.read_csv('modified_data/movie_metadata_TMDB.csv')
print(f"Added TMDB ID to {df['TMDB_id'].count()} movies. Total movies: {df['Wikipedia Movie ID'].count()}. \n Percentage populated: {round(df['TMDB_id'].count()/df['Wikipedia Movie ID'].count()*100,2)}%")


Added TMDB ID to 68944 movies. Total movies: 81741. 
 Percentage populated: 84.34%


2 - Adding Movie Revenue
Since the script above (enrich_v1) of the data collection used a 'query' method for TMDB, we are now going to be using the 'Movies' method to collect further data about the movies (enrich_v2). 
Since many of the movies are missing their revenue, TMDB might have that information for some movies.

The same basis as for enrich_v1 is used to collect data. The data collected is: 
- revenue, budget, runtime, IMDB-id and genres.

While our original dataset includes movie genres, it's always good to have more information and the two might differ for an interesting analysis further on.

Like enrich_v1, enrich_v2 was run separately in another file: /scratch/add_TMDB_movie_metadata_V2.py as it needed to be run on several occasions since the API is limited.

In [ ]:
import tmdbsimple as tmdb
from tqdm import tqdm
import pandas as pd
from datetime import datetime
import re
from dotenv import load_dotenv
import os

load_dotenv()

# Load v1, which was created in part 1 above
movie_metadata_TMDB = pd.read_csv('modified_data/movie_metadata_TMDB.csv')
movie_metadata_TMDB = movie_metadata_TMDB.dropna(subset=['TMDB_id'])

# Load API key
TMDB_API_KEY = os.environ.get("TMDB_API_KEY")
tmdb.API_KEY = TMDB_API_KEY
tmdb.REQUESTS_TIMEOUT = 5  # seconds, for both connect and read

# Create a list to save progress
saved_progress = []
# Determine where to resume
start_index = 35000

# Progress file that can be used to resume
#saved_progress = pd.read_json('progress_v2.json')['index'].tolist()
#start_index = saved_progress[-1] + 1  # Start from the next index

for index, row in tqdm(movie_metadata_TMDB.iterrows(), total=len(movie_metadata_TMDB), desc="Processing"): # Wraps for loop in progress bar
    if index < start_index:
        continue
    try:
        movie = tmdb.Movies(row['TMDB_id']) # The "Movies" method is used with the TMDB_id of the movie
        movie_info = movie.info()
        
        # The dataframe is populated with data about: runtime, revenue, budget and IMDB_id.
        movie_metadata_TMDB.loc[index,'TMDB_runtime'] = movie_info['runtime']
        movie_metadata_TMDB.loc[index,'TMDB_revenue'] = movie_info['revenue']
        movie_metadata_TMDB.loc[index,'TMDB_budget'] = movie_info['budget']
        movie_metadata_TMDB.loc[index,'TMDB_IMDB_id'] = movie_info['imdb_id']

        # Convert list of genres to a string representation for adding the values easily to the df
        genres_str = ', '.join([genre['name'] for genre in movie_info['genres']])
        # Assign values to DataFrame
        movie_metadata_TMDB.loc[index, 'TMDB_genres'] = genres_str
        
        # Save the index as progress
        saved_progress.append(index)

        # Save progress periodically (in case of interruption)
        if index % 50 == 0:
            progress_df = pd.DataFrame({'index': saved_progress})
            progress_df.to_json('progress_v2.json')
            movie_metadata_TMDB.to_csv('movie_metadata_TMDB_v2.csv', index=False)

    except Exception as e:
        print(f"Error at index {index}: {str(e)}")

# Save final progress
progress_df = pd.DataFrame({'index': saved_progress})
progress_df.to_json('progress_v2.json')

# Save your final DataFrame
movie_metadata_TMDB.to_csv('movie_metadata_TMDB_v2_DONE.csv', index=False)


The data collected is now analysed below.

In [34]:
import numpy as np
import pandas as pd

movie_metadata_v2 = pd.read_csv('modified_data/movie_metadata_TMDB_v2.csv')
original_revenue = movie_metadata_v2['Movie box office revenue'].isnull().sum() # Calculate the total NaN values are in the provided datasets revenue
tmdb_v2_revenue = movie_metadata_v2['TMDB_revenue'].isnull().sum() # Calculate total NaN values for v2

print("Original revenue NaN:", original_revenue)
print("New revenue NaN:",tmdb_v2_revenue)
print("Difference", original_revenue-tmdb_v2_revenue)

display(movie_metadata_v2)

Original revenue NaN: 60666
New revenue NaN: 59965
Difference 701


Unnamed: 0.1  Unnamed: 0  Wikipedia Movie ID Freebase Movie ID  \
0                 0           0              975900         /m/03vyhn   
1                 1           1             3196793         /m/08yl5d   
2                 2           2            28463795        /m/0crgdbh   
3                 3           3             9363483        /m/0285_cd   
4                 4           4              261236         /m/01mrr1   
...             ...         ...                 ...               ...   
68939         68939       81735            32468537        /m/0crwd9y   
68940         68940       81736            35228177        /m/0j7hxnt   
68941         68941       81737            34980460        /m/0g4pl34   
68942         68942       81738             9971909        /m/02pygw1   
68943         68943       81740            12476867        /m/02w7zz8   

                                              Movie name Movie release date  \
0                                         Ghosts of Mars         2001-08-24   
1      Getting Away with Murder: The JonBenét Ramsey ...         2000-02-16   
2                                            Brun bitter               1988   
3                                       White Of The Eye               1987   
4                                      A Woman in Flames               1983   
...                                                  ...                ...   
68939                                    Shadow Boxing 2         2007-10-18   
68940                           Mermaids: The Body Found         2011-03-19   
68941                                            Knuckle         2011-01-21   
68942                                  Another Nice Mess         1972-09-22   
68943                                            Spliced               2002   

       Movie box office revenue  Movie runtime  \
0                    14010832.0           98.0   
1                           NaN           95.0   
2                           NaN           83.0   
3                           NaN          110.0   
4                           NaN          106.0   
...                         ...            ...   
68939                       NaN          132.0   
68940                       NaN          120.0   
68941                       NaN           96.0   
68942                       NaN           66.0   
68943                       NaN           86.0   

                                         Movie languages  \
0                     {"/m/02h40lc": "English Language"}   
1                     {"/m/02h40lc": "English Language"}   
2                     {"/m/05f_3": "Norwegian Language"}   
3                     {"/m/02h40lc": "English Language"}   
4                      {"/m/04306rv": "German Language"}   
...                                                  ...   
68939  {"/m/06b_j": "Russian Language", "/m/02h40lc":...   
68940                 {"/m/02h40lc": "English Language"}   
68941                 {"/m/02h40lc": "English Language"}   
68942                 {"/m/02h40lc": "English Language"}   
68943                 {"/m/02h40lc": "English Language"}   

                                         Movie countries  ... TMDB_popularity  \
0              {"/m/09c7w0": "United States of America"}  ...          17.280   
1              {"/m/09c7w0": "United States of America"}  ...           0.750   
2                                 {"/m/05b4w": "Norway"}  ...           0.600   
3                         {"/m/07ssc": "United Kingdom"}  ...           7.336   
4                                {"/m/0345h": "Germany"}  ...           2.397   
...                                                  ...  ...             ...   
68939                             {"/m/06bnz": "Russia"}  ...           4.421   
68940          {"/m/09c7w0": "United States of America"}  ...           5.098   
68941  {"/m/03rt9": "Ireland", "/m/07ssc": "United Ki...  ...           9.789   
68942          {"/m/09c7w0": "United States of Am

It seems like the original dataset contained much of the information available about the movie revenue.
However, for 701 movies the revenue was added, which is quite a lot still in the context of ~40000 movies with plot summaries.



2.1 - combining revenue columns into a single one.
In this section we're going to combine the original movie revenue data with the collected data in enrich_v2.


In [35]:
print('OLD rev',movie_metadata_v2['Movie box office revenue'].isnull().sum())
print('NEW rev:',movie_metadata_v2['TMDB_revenue'].isnull().sum())

movie_metadata_v2['Movie box office revenue enriched'] = movie_metadata_v2['Movie box office revenue']
movie_metadata_v2['Movie box office revenue enriched'].fillna(movie_metadata_v2['TMDB_revenue'])
#movie_metadata_v2.drop(columns=['TMDB_revenue','Movie box office revenue'])

display(movie_metadata_v2)




OLD rev 60666
NEW rev: 59965


Unnamed: 0.1  Unnamed: 0  Wikipedia Movie ID Freebase Movie ID  \
0                 0           0              975900         /m/03vyhn   
1                 1           1             3196793         /m/08yl5d   
2                 2           2            28463795        /m/0crgdbh   
3                 3           3             9363483        /m/0285_cd   
4                 4           4              261236         /m/01mrr1   
...             ...         ...                 ...               ...   
68939         68939       81735            32468537        /m/0crwd9y   
68940         68940       81736            35228177        /m/0j7hxnt   
68941         68941       81737            34980460        /m/0g4pl34   
68942         68942       81738             9971909        /m/02pygw1   
68943         68943       81740            12476867        /m/02w7zz8   

                                              Movie name Movie release date  \
0                                         Ghosts of Mars         2001-08-24   
1      Getting Away with Murder: The JonBenét Ramsey ...         2000-02-16   
2                                            Brun bitter               1988   
3                                       White Of The Eye               1987   
4                                      A Woman in Flames               1983   
...                                                  ...                ...   
68939                                    Shadow Boxing 2         2007-10-18   
68940                           Mermaids: The Body Found         2011-03-19   
68941                                            Knuckle         2011-01-21   
68942                                  Another Nice Mess         1972-09-22   
68943                                            Spliced               2002   

       Movie box office revenue  Movie runtime  \
0                    14010832.0           98.0   
1                           NaN           95.0   
2                           NaN           83.0   
3                           NaN          110.0   
4                           NaN          106.0   
...                         ...            ...   
68939                       NaN          132.0   
68940                       NaN          120.0   
68941                       NaN           96.0   
68942                       NaN           66.0   
68943                       NaN           86.0   

                                         Movie languages  \
0                     {"/m/02h40lc": "English Language"}   
1                     {"/m/02h40lc": "English Language"}   
2                     {"/m/05f_3": "Norwegian Language"}   
3                     {"/m/02h40lc": "English Language"}   
4                      {"/m/04306rv": "German Language"}   
...                                                  ...   
68939  {"/m/06b_j": "Russian Language", "/m/02h40lc":...   
68940                 {"/m/02h40lc": "English Language"}   
68941                 {"/m/02h40lc": "English Language"}   
68942                 {"/m/02h40lc": "English Language"}   
68943                 {"/m/02h40lc": "English Language"}   

                                         Movie countries  ...  \
0              {"/m/09c7w0": "United States of America"}  ...   
1              {"/m/09c7w0": "United States of America"}  ...   
2                                 {"/m/05b4w": "Norway"}  ...   
3                         {"/m/07ssc": "United Kingdom"}  ...   
4                                {"/m/0345h": "Germany"}  ...   
...                                                  ...  ...   
68939                             {"/m/06bnz": "Russia"}  ...   
68940          {"/m/09c7w0": "United States of America"}  ...   
68941  {"/m/03rt9": "Ireland", "/m/07ssc": "United Ki...  ...   
68942          {"/m/09c7w0": "United States of America"}  ...   
68943                            {"/m/0d060g": "Canada"}  ...   

      TMDB_release_date                                         TMDB_title  \
